cluster-autoscaler/cloudprovider/hetzner/hetzner_servers_cache.go

```diff
@@ -136,7 +135,8 @@ func (m *serversCache) getServer(nodeIdOrName string) (*hcloud.Server, error) {
 		}
 	}
 
-	return nil, errors.New("server not found")
+	// return nil if server not found
+	return nil, nil
```

Comment:
This "fixes" two issues:

1. Deleted servers cause issues until the Kubernetes Node is deleted too
2. Servers with different Provider IDs than `hcloud://...` in the cluster break cluster-autoscaler

I think this is a good fix for the first "bug".

For the second bug, I think we should make this more explicit and handle it where we "parse" the provider ID. What do you think about this patch?

```patch
diff --git a/cluster-autoscaler/cloudprovider/hetzner/hetzner_manager.go b/cluster-autoscaler/cloudprovider/hetzner/hetzner_manager.go
index f75c69ee3..35a256f46 100644
--- a/cluster-autoscaler/cloudprovider/hetzner/hetzner_manager.go
+++ b/cluster-autoscaler/cloudprovider/hetzner/hetzner_manager.go
@@ -259,6 +259,12 @@ func (m *hetznerManager) addNodeToDrainingPool(node *apiv1.Node) (*hetznerNodeGr
 func (m *hetznerManager) serverForNode(node *apiv1.Node) (*hcloud.Server, error) {
        var nodeIdOrName string
        if node.Spec.ProviderID != "" {
+               if !strings.HasPrefix(node.Spec.ProviderID, providerIDPrefix) {
+                       // This cluster-autoscaler provider only handles Hetzner Cloud servers.
+                       // Any other provider ID prefix is invalid, and we return no server. Returning an error here breaks hybrid
+                       // clusters with nodes from Hetzner Cloud & Robot (or other providers).
+                       return nil, nil
+               }
                nodeIdOrName = strings.TrimPrefix(node.Spec.ProviderID, providerIDPrefix)
        } else {
                nodeIdOrName = node.Name
```

Reply:
That makes sense. Thanks.